File: backend/src/main/scala/services/battle/microservices/combat/objects/projectile/BattleProjectileState.scala
package services.battle.microservices.combat.objects.projectile

import services.battle.microservices.combat.objects.projectile.{ProjectileKind, ProjectileTerminalReason}
import services.battle.microservices.combat.objects.combat.Damage
import services.battle.microservices.actors.objects.player.HitPoints
import services.battle.objects.core.{
  BattleVector2,
  DurationMillis,
  ElapsedMillis,
  FacingRadians,
  HeroId,
  PlayerId,
  ProjectileId,
  Radius
}

final case class BattleProjectileState(
  projectileId: ProjectileId,
  ownerHeroId: HeroId,
  projectileKind: ProjectileKind,
  position: BattleVector2,
  velocity: BattleVector2,
  facing: FacingRadians,
  radius: Radius,
  damage: Damage,
  ttlMs: DurationMillis,
  maxLifetimeMs: DurationMillis,
  splashRadius: Radius
)

final case class BattleProjectileTerminalState(
  projectileId: ProjectileId,
  projectileKind: ProjectileKind,
  ownerPlayerId: PlayerId,
  ownerHeroId: HeroId,
  reason: ProjectileTerminalReason,
  start: BattleVector2,
  end: BattleVector2,
  terminalPosition: BattleVector2,
  ttlBefore: DurationMillis,
  ttlAfter: DurationMillis,
  elapsedMs: ElapsedMillis,
  targetPlayerId: Option[PlayerId],
  targetHeroId: Option[HeroId],
  hpBefore: Option[HitPoints],
  hpAfter: Option[HitPoints],
  damage: Option[Damage]
)

File: frontend/src/objects/battle/microservices/combat/objects/projectile/BattleProjectileState.ts
import type { BattleTeamMode, BattleVector2 } from "../../../../objects/core/BattleCoreScalars";
import type { Damage } from "../combat/Damage";
import type { ProjectileKind } from "./ProjectileKind";
import type { ProjectileTerminalReason } from "./ProjectileTerminalReason";

export interface BattleProjectileState {
  projectileId: string;
  ownerHeroId: string;
  kind: ProjectileKind;
  team: BattleTeamMode;
  position: BattleVector2;
  velocity: BattleVector2;
  facing: number;
  radius: number;
  damage: Damage;
  ttlMs: number;
  maxLifetimeMs: number;
  splashRadius: number;
  alive: boolean;
  hitTargets: string[];
}

export interface BattleProjectileTerminalState {
  projectileId: string;
  kind: ProjectileKind;
  ownerPlayerId: string;
  ownerHeroId: string;
  reason: ProjectileTerminalReason;
  start: BattleVector2;
  end: BattleVector2;
  terminalPosition: BattleVector2;
  ttlBefore: number;
  ttlAfter: number;
  elapsedMs: number;
  targetPlayerId: string | null;
  targetHeroId: string | null;
  hpBefore: number | null;
  hpAfter: number | null;
  damage: Damage | null;
}
